### ISSUE TITLE
Child actions don't dispatch to stores if Parent actions are async

### ISSUE THREAD

**Issue by ericclemmons** [acdlite/flummox#122]
So, @willrstern & I have spent a lot of time debugging this and having a functional example to prove our case:
> https://github.com/ericclemmons/flummox-parent-promise

So, there's a couple things at play here:

- We have a custom `FluxContainer` HoC that:
  + Runs `actions` on `componentWillMount` (so that stores can get the data they need)
  + Returns a `<FluxComponent connectToStores={...} render={...} />` whose `render` function ensures
     the original component renders only when all of the store getters return a value.

(This is to ensure that components that have `React.PropTypes...isRequired` don't error out unnecessarily)

When running this demo, we noticed several bugs:

- Nothing seems to render at all (perhaps because of #60?)
- The `ChildStore` never gets notified when `ChildActions#fetch` completes, despite
  `ParentStore` getting notified when `ParentActions#fetch` completes.

**Suffice it to say, this is a show-stopper for us, so if you'd like, you can hit me up on *Reactiflux* and we can work out a bounty, @acdlite!**.

**Comment by ericclemmons** [comment]
Also:

- Does it matter that we have `myAction() { return new Promise(...) }` vs. `async myAction() { return new Promise(...) }`?

**Comment by acdlite** [comment]
Nope, no difference. Just needs to return a promise.

**Comment by acdlite** [comment]
Okay, so this is happening due to a combination of two problems. First is our favorite Dispatcher error:

```js
"Invariant Violation: Dispatch.dispatch(...): Cannot dispatch in the middle of a dispatch."
```

There are a few ways to fix that (including [overriding the default Dispatcher](https://github.com/acdlite/flummox/pull/107)) but the real problem is...

...the error was being swallowed! (by https://github.com/acdlite/flummox/blob/master/src/Actions.js#L57-L59) This same issue has caused a few other problems. I will push a fix very soon (tonight).

**Comment by ericclemmons** [comment]
Interesting!  So this may mean something else entirely for me (us?)...

If sub-action dispatching poses a problem, it may mean we have to re-architect our stores to not be interdependent but rely on _Services_ instead.

Remember how I mentioned before that we're doing sub-dispatching?  Here's the flow that our forms take:

- Client loads up a form to edit a user
  + `UserFormActions#create`
    - `UserActions#create`
    - `UserActions#create`'s return value is ran through a series of promises to make it into a useful view model
  + `UserFormStore#onCreate`
- Dispatch occurs & the view renders with a `form` based on a `user` model.
- User edits field, firing off `onChange` within the component
  + `UserFormActions#update(form, property, value)`
    - `UserActions#update(user, property, value)` modifies the model with normalized values & computes any other properties
    - Returns the `user` with new values
  + `UserFormStore#onUpdate(user)` re-generates the appropriate parts of the view model
- Updated form is set on `setState` and the view re-renders.

You can see this inter-dependency with Actions/Stores because:

- The `UserForm*` classes are pretty much for reducing & mapping a `User` model into a form's view model.
- The `User*` classes are the true owners of domain-logic.
- On the server **no forms are required** to create a valid, computed, normalized `User` model.

**Comment by ericclemmons** [comment]
Closing.

- Error swallowing bug fixed in `3.3.1`
- Dispatch within a dispatch fixed via:
  + https://github.com/ericclemmons/flummox-parent-promise/pull/2
